aten/src/ATen/native/Distance.cpp

```diff
@@ -159,13 +159,46 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
 }
 
 Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double eps) {
-  // Follow scipy impl to improve numerical precision
-  // Use x / sqrt(x * x) instead of x / (sqrt(x) * sqrt(x))
-  Tensor w12 = at::sum(x1 * x2, dim);
-  Tensor w1 = at::sum(x1 * x1, dim);
-  Tensor w2 = at::sum(x2 * x2, dim);
-  Tensor n12 = (w1 * w2).clamp_min_(eps * eps).sqrt_();
-  return w12.div_(n12);
+  /*
+   * cosine_similarity(x1, x2) = <x1, x2> / (||x1|| * ||x2||)
+   * 
+   * The current implementation is an improvement over the previous version.
+   * 
+   * Previous implementation:
+   * 1. Compute num = <x1, x2>,
+   * 2. Compute denom = ||x1|| * ||x2||,
+   * 3. Compute denom = max(denom, eps) to avoid division by zero,
+   * 4. Return num / denom.
+   * 
+   * Previous implementation has the following issues:
+   * 1. Chance of loosing precision in <x1, x2> when ||x1|| and ||x2|| are large.
+   * 2. Chance of loosing precision in ||x1|| * ||x2|| when ||x1|| and ||x2|| are large.
+   * 3. Loosing precision may cause |cosing_similarity(x1, x2)| > 1.0.
+   * 
+   * Current implementation:
+   * 1. Compute x1_normalized = x1 / (||x1|| + eps),
```

Comment:
The change from `max(foo, eps)` to ` foo + eps` is to enforce that the normalized vectors have norms <1 ?
The goal being that floating point errors in the dot product won't push us >1 ?
How big should eps be to make sure this is always true?

Reply:
It is to enforce that we do not divide by zero, and yes, as a side effect, it makes norms less than 1. I can't say I am 100% sure that this will work for any epsilon, but it does work for the standard epsilon (1e-8). Anyway, this implementation should tolerate larger set of epsilon values compared to the older algorithm because the vectors in the inner product live in a unit ball, and mantissa of floating point values is utilized much better.

Reply:
But the difference with the `max` is that if the norm is small (like `2*eps`), you will make a much bigger error here. Is that something we should worry about?

Reply:
Well, that is a philosophical dilemma of how to handle zeroes. I personally do not like using max, because it makes function non-differentiable in a non-zero measure set of epsilon ball...

Reply:
Maybe you are right and it makes sense to do some case distinction...

Reply:
You can avoid this problem by doing this change in a non differentiable manner using `NoGradGuard` around the `max` operations. That way, this won't change the gradient computation.

Reply:
Could you please explain why then clapping did not introduce any issues in the older implementation?

Reply:
The current implementation do not have it.
And the other PR should have the `NoGradGuard` as well to avoid this issue. There is a comment in it asking to do so.